### ISSUE TITLE
error: cadvisor.go:338: Unable to perform lookup under /sys/dev/ for block device major(nnn) minor(nn)

### ISSUE THREAD

**Issue by gbrayut** [bosun-monitor/bosun#1869]
It looks like #1648 changed the cadvisor collector so that it generates errors if you try to monitor a remote endpoint. I added some print statements to see what it is doing:

```
C:\Code\Go\src\bosun.org\cmd\scollector [master +0 ~1 -0]> go build .;.\scollector -f cadvisor
2016/08/12 18:22:33 info: cadvisor.go:530: startCadvisorCollector
2016/08/12 18:22:33 info: cadvisor.go:532: startCadvisorCollector: http://ny-gbraylx01.ds.stackexchange.com:8080
2016/08/12 18:22:33 info: main.go:217: OpenTSDB host: http://ny-gbraylx01:8070
2016/08/12 18:22:33 info: cadvisor.go:514: c_cadvisor 1
2016/08/12 18:22:33 info: cadvisor.go:522: c_cadvisor range container: /docker/7472a3a1694b642fa885de51fa05d3c00056bf52ecebfcc6400acdd0e709343b
2016/08/12 18:22:33 info: cadvisor.go:374: statsForContainer
2016/08/12 18:22:33 info: cadvisor.go:395: statsForContainer HasFilesystem
2016/08/12 18:22:33 info: cadvisor.go:416: statsForContainer HasMemory
2016/08/12 18:22:33 info: cadvisor.go:430: statsForContainer HasNetwork
2016/08/12 18:22:33 info: cadvisor.go:469: statsForContainer HasDiskIo (errors here)
2016/08/12 18:22:33 info: cadvisor.go:328: addBlkioStat
2016/08/12 18:22:33 error: cadvisor.go:341: Unable to perform lookup under /sys/dev/ for block device major(7) minor(0)
2016/08/12 18:22:33 info: cadvisor.go:330: addBlkioStat device: major7_minor0
2016/08/12 18:22:33 info: cadvisor.go:332: addBlkioStat label: Read
2016/08/12 18:22:33 info: cadvisor.go:332: addBlkioStat label: Sync
2016/08/12 18:22:33 info: cadvisor.go:332: addBlkioStat label: Total
2016/08/12 18:22:33 info: cadvisor.go:332: addBlkioStat label: Write
2016/08/12 18:22:33 info: cadvisor.go:332: addBlkioStat label: Async
2016/08/12 18:22:33 info: cadvisor.go:328: addBlkioStat
... repeats 5 times ...
2016/08/12 18:22:33 info: cadvisor.go:473: statsForContainer HasDiskIo 2  (errors here)
... similar errors here repeated 5 times ...
2016/08/12 18:22:33 info: cadvisor.go:477: statsForContainer HasDiskIo 3
2016/08/12 18:22:33 info: cadvisor.go:482: statsForContainer HasDiskIo 4
2016/08/12 18:22:33 info: cadvisor.go:487: statsForContainer HasDiskIo 5
2016/08/12 18:22:33 info: cadvisor.go:492: statsForContainer HasDiskIo 6
2016/08/12 18:22:33 info: cadvisor.go:497: statsForContainer HasDiskIo 7
2016/08/12 18:22:33 info: cadvisor.go:502: statsForContainer HasDiskIo 8
2016/08/12 18:22:33 info: cadvisor.go:507: statsForContainer HasDiskIo 9
2016/08/12 18:22:33 info: cadvisor.go:509: statsForContainer finished
```

So it looks like the range over [stats.DiskIo.IoServiceBytes](https://github.com/bosun-monitor/bosun/blob/ac7f667fa1e649b877b77bfab7f2a643f5d8e615/cmd/scollector/collectors/cadvisor.go#L462) and stats.DiskIo.IoServiced are generating errors in our logs because we monitor multiple cadvisor endpoints from a single remote scollector instance.

That means the [lookup in the local /sys/dev/block/%d:%d/uevent](https://github.com/bosun-monitor/bosun/blob/ac7f667fa1e649b877b77bfab7f2a643f5d8e615/cmd/scollector/collectors/cadvisor.go#L342) is going to fail. Instead we should use the information from the disk_map section of the http://endpoint:port/api/v1.3/machine to map the ids to their names. Example of Raw output:

```
{
	"num_cores": 2,
	"cpu_frequency_khz": 2299998,
	"memory_capacity": 8233426944,
	"machine_id": "f7ba7e5d525340a2b12c76afafad3f78",
	"system_uuid": "421D89A1-CDA9-171B-89C2-3C3C903EBD64",
	"boot_id": "ac56431e-0ff6-418d-8ab1-8925dd86d918",
	"filesystems": [{
		"device": "/dev/mapper/docker-253:0-419445-9232c1b180a363da253655efcc6c88021405b0a47bd1092bb69c9c69d82a51ac",
		"capacity": 105554829312,
		"type": "vfs",
		"inodes": 6553600
	},
	{
		"device": "/dev/mapper/primary-vg_root",
		"capacity": 38040530944,
		"type": "vfs",
		"inodes": 37167104
	},
	{
		"device": "/dev/sda1",
		"capacity": 545959936,
		"type": "vfs",
		"inodes": 536576
	},
	{
		"device": "/dev/mapper/docker-253:0-419445-7472a3a1694b642fa885de51fa05d3c00056bf52ecebfcc6400acdd0e709343b",
		"capacity": 105554829312,
		"type": "vfs",
		"inodes": 6553600
	}],
	"disk_map": {
		"253:0": {
			"name": "dm-0",
			"major": 253,
			"minor": 0,
			"size": 38059114496,
			"scheduler": "none"
		},
		.... omitted additional lines here ...
		"8:0": {
			"name": "sda",
			"major": 8,
			"minor": 0,
			"size": 42949672960,
			"scheduler": "deadline"
		}
	},
	...
}
```

That way we can still support polling of remote endpoints. @mathpl do you have any objections to this approach? If not I'm happy to create a PR using the the [client.MachineInfo()](https://github.com/google/cadvisor/blob/6213bfab8daa8e06960a227c3f6ece266b55bd1a/info/v1/machine.go#L170) disk_map instead of direct calls to /sys/dev/block/%d:%d/uevent.

**Comment by gbrayut** [comment]
I created a PR at https://github.com/bosun-monitor/bosun/pull/1870 for discussion.

**Comment by gbrayut** [comment]
fixed in e8080a36843b6effbf521ab7e96fcdc3acc3a220 using isRemote flag
